spec/models/favorite_tag_spec.rb

```diff
@@ -0,0 +1,49 @@
+require 'spec_helper'
+
+describe FavoriteTag do
+  let(:noncanonical_favorite) { build(:favorite_tag,
+                                      tag_id: create(:freeform).id) }
+  let(:no_tag_favorite) { build(:favorite_tag, tag_id: nil) }
+  let(:no_user_favorite) { build(:favorite_tag, user_id: nil) }
+
+  it "has a valid factory" do
+    expect(create(:favorite_tag)).to be_valid
+  end
+
+  it "is invalid without a tag_id" do
+    expect(no_tag_favorite.save).to be_falsey
+  end
+
+  it "is invalid when the user has already favorited the tag" do
+    user = create(:user)
+    tag = create(:canonical_freeform)
+    create(:favorite_tag, tag_id: tag.id, user_id: user.id)
+    expect(build(:favorite_tag, tag_id: tag.id, user_id: user.id).save).to be_falsey
+  end
+
+  it "is invalid without a user_id" do
+    expect(no_user_favorite.save).to be_falsey
+  end
+
+  it "is invalid with a non-canonical tag" do
+    expect(noncanonical_favorite.save).to be_falsey
```

Comment:
Some of these `save` calls can be replaced with `valid?`. We can test validations without saving to the database which makes it faster.